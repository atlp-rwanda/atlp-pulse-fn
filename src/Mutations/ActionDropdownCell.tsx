import React from 'react';
import ActionDropdown from '../components/ActionDropdown';

interface ActionDropdownCellProps {
  row: {
    original: any;
  };
  onView: (ticket: any) => void;
  onEdit?: (id: string) => void;
  onDelete?: (id: string) => void;
  canEditDelete: boolean;
}

function ActionDropdownCell({
  row,
  onView,
  onEdit,
  onDelete,
  canEditDelete,
}: ActionDropdownCellProps) {
  return (
    <ActionDropdown
      onView={() => onView(row.original)}
      onEdit={canEditDelete ? () => onEdit?.(row.original.id) : undefined}
      onDelete={canEditDelete ? () => onDelete?.(row.original.id) : undefined}
      canEditDelete={canEditDelete}
      id={row.original.id}
    />
  );
}

export default ActionDropdownCell;
