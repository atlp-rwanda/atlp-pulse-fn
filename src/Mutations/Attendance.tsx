import {gql} from '@apollo/client';

export const GET_TEAM_ATTENDANCE = gql`
  query GetTeamAttendance($team: String!, $orgToken: String) {
    getTeamAttendance(team: $team, orgToken: $orgToken) {
      id
      week
      phase {
        id
        name
      }
      cohort {
        id
        name
      }
      teams {
        team {
          id
          name
        }
        trainees {
          trainee {
            id
            email
            profile {
              id
              name
            }
          }
          status {
            day
            date
            score
          }
        }
      }
    }
  }
`;
export const GET_ATTENDANCE_BY_ID = gql`
  query GetAttendance($id: ID!) {
    getTraineeAttendanceByID(id: $id) {
      id
      week
      trainees {
        traineeId
        traineeEmail
        status {
          days
          value
        }
      }
    }
  }
`;
export const GET_WEEKLY_ATTENDANCE = gql`
  query GetTraineeAttendanceByID($traineeEmail: String!) {
    getTraineeAttendanceByID(traineeEmail: $traineeEmail) {
      weekNumber
      traineeAttendance {
        days
        value
      }
    }
  }
`;

export const RECORD_ATTENDANCE = gql`
  mutation RecordAttendance(
    $week: Int!
    $team: String!
    $date: String!
    $trainees: [TraineeInput!]!
    $orgToken: String!
  ) {
    recordAttendance(
      week: $week
      team: $team
      date: $date
      trainees: $trainees
      orgToken: $orgToken
    ) {
      team {
        id
        name
        cohort {
          name
        }
      }
      trainees {
        trainee {
          profile {
            name
          }
        }
        status {
          day
          date
          score
        }
      }
    }
  }
`;

export const UPDATE_ATTENDANCE = gql`
  mutation UpdateAttendance(
    $week: Int!
    $team: String!
    $phase: String!
    $trainees: [TraineeInput!]!
    $orgToken: String!
  ) {
    updateAttendance(
      week: $week
      team: $team
      phase: $phase
      trainees: $trainees
      orgToken: $orgToken
    ) {
      id
      week
      phase {
        id
        name
      }
      cohort {
        id
        name
      }
      teams {
        team {
          id
          name
        }
        trainees {
          trainee {
            id
            email
            profile {
              id
              name
            }
          }
          status {
            day
            date
            score
          }
        }
      }
    }
  }
`;

export const DELETE_ATTENDANCE = gql`
  mutation DeleteAttendance($week: String!, $team: String!, $day: String!) {
    deleteAttendance(week: $week, team: $team, day: $day) {
      id
      week
      phase {
        id
        name
      }
      cohort {
        id
        name
      }
      teams {
        team {
          id
          name
        }
        trainees {
          trainee {
            id
            email
            profile {
              id
              name
            }
          }
          status {
            day
            date
            score
          }
        }
      }
    }
  }
`;
