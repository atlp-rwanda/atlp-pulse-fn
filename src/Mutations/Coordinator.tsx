import { gql } from '@apollo/client';


export const GetAllCoordinatorUsers = gql`
  query GetCoordinatorUsers($orgToken: String) {
    getAllCoordinatorUsers(orgToken: $orgToken) {
      id
      email
      profile {
        name
      }
    }
  }
  
`;