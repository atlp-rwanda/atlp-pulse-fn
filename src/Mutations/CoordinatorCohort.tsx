import { gql } from '@apollo/client';

export const GetCoordinatorCohort = gql`
  query GetCoordinatorCohort($coordinatorId: String) {
    getCoordinatorCohort(coordinator_id: $coordinatorId) {
      program {
        name
      }
      name
    }
  }
`;
