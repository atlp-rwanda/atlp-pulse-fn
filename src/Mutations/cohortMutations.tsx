import { gql } from '@apollo/client';

export const AddCohort = gql`
  mutation AddCohort(
    $name: String!
    $phaseName: String!
    $coordinatorEmail: String!
    $programName: String!
    $startDate: DateTime!
    $endDate: DateTime
    $orgToken: String!
  ) {
    addCohort(
      name: $name
      phaseName: $phaseName
      coordinatorEmail: $coordinatorEmail
      programName: $programName
      startDate: $startDate
      endDate: $endDate
      orgToken: $orgToken
    ) {
      id
      name
      phase {
        name
      }
      coordinator {
        email
      }
      program {
        name
      }
      startDate
      endDate
    }
  }
`;

export const DeleteCohort = gql`
  mutation DeleteCohort($deleteCohortId: ID!, $orgToken: String) {
    deleteCohort(id: $deleteCohortId, orgToken: $orgToken) {
      id
    }
  }
`;
