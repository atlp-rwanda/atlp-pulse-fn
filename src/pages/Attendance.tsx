import React from 'react';
import CheckRole from '../utils/CheckRoles';

const TraineeAttendanceTracker = React.lazy(
  () => import('../pages/TraineeAttendance'),
);
const TraineeAttendance = React.lazy(
  () => import('../components/TraineeAttendance'),
);

function Attendance() {
  return (
    <>
      <CheckRole roles={['coordinator']}>
        <TraineeAttendanceTracker />
      </CheckRole>
      <CheckRole roles={['trainee']}>
        <TraineeAttendance />
      </CheckRole>
    </>
  );
}

export default Attendance;
