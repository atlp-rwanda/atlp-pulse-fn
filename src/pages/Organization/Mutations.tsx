import { gql } from '@apollo/client';

export const SIGN_UP_MUTATION = gql`
  mutation Mutation(
    $firstName: String!
    $lastName: String!
    $dateOfBirth: DateTime!
    $gender: String!
    $email: String!
    $password: String!
    $orgToken: String!
  ) {
    createUser(
      firstName: $firstName
      lastName: $lastName
      dateOfBirth: $dateOfBirth
      gender: $gender
      email: $email
      password: $password
      orgToken: $orgToken
    ) {
      user {
        profile {
          name
        }
      }
    }
  }
`;

export const GET_SIGNUP_ORGANIZATION = gql`
  query GetSignupOrganization($orgToken: String!) {
    getSignupOrganization(orgToken: $orgToken) {
      name
    }
  }
`;
