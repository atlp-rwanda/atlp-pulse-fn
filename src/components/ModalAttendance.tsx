import React, { useEffect, useState } from 'react';
import { toast } from 'react-toastify';
import { useMutation } from '@apollo/client';
import { format } from 'date-fns';
import { RECORD_ATTENDANCE } from '../Mutations/Attendance';
import AttendanceSymbols from './AttendanceSymbols';

interface ModalProps {
  isVisible: boolean;
  onClose: () => void;
  setAttendanceData: React.Dispatch<React.SetStateAction<any[]>>;
  trainees: any;
  week: number;
  date: string;
  day: string;
  team: string;
  teamName: string;
}
interface attendanceProps {
  name: string;
  score: number;
  id: string;
  day: string;
}

interface StatusInput {
  day: string;
  score: string;
}

export interface recordTraineeProps {
  trainee: string;
  status: StatusInput;
}

function ModalAttendance({
  isVisible,
  onClose,
  trainees,
  week,
  date,
  day,
  setAttendanceData,
  team,
  teamName,
}: ModalProps) {
  const [allTrainees, setAllTrainees] = useState<any[]>([]);
  const [filteredTrainees, setFilteredTrainees] = useState<any[]>([]);
  const [searchName, setSearchName] = useState('');
  const [inputFocus, setInputFocus] = useState(false);
  const [traineesAttendance, setTraineesAttendance] = useState<
    attendanceProps[]
  >([]);
  const [recordTrainees, setRecordTrainees] = useState<recordTraineeProps[]>(
    [],
  );

  const [
    recordAttendance,
    { data: recordAttendanceData, loading: loadingRecordAttendance, error },
  ] = useMutation(RECORD_ATTENDANCE, {
    variables: {
      week,
      team,
      date,
      trainees: recordTrainees,
      orgToken: localStorage.getItem('orgToken'),
    },
    fetchPolicy: 'no-cache',
    onCompleted: (data) => {
      if (data) {
        toast.success('Attendance recorded successfully.');
      }

      setAttendanceData((prev) => {
        let isSet = false;
        const result = prev.map((attendance) => {
          if (
            attendance.week.toString() === week.toString() &&
            attendance.cohort.id === data.recordAttendance.team.cohort.id &&
            attendance.phase.id === data.recordAttendance.team.cohort.phase.id
          ) {
            isSet = true;
            return {
              ...attendance,
              teams: [data.recordAttendance],
            };
          }
          return attendance;
        });

        if (!isSet && team === data.recordAttendance.team.id) {
          result.push({
            week: week.toString(),
            id: Math.random().toString(),
            cohort: {
              id: data.recordAttendance.team.cohort.id,
              name: data.recordAttendance.team.cohort.name,
            },
            phase: {
              id: data.recordAttendance.team.cohort.phase.id,
              name: data.recordAttendance.team.cohort.phase.name,
            },
            teams: [data.recordAttendance],
          });
        }
        return result;
      });
      setTraineesAttendance([]);
      setRecordTrainees([]);
      onClose();
    },
    onError: (error) => {
      const errorMessage =
        error.graphQLErrors?.[0]?.message || 'An unexpected error occurred';
      toast.error(errorMessage);
    },
  });

  useEffect(() => {
    if (recordTrainees.length > 0) {
      recordAttendance();
    }
  }, [recordTrainees]);

  useEffect(() => {
    if (trainees) {
      const sanitizedTrainees = trainees.filter(
        (trainee: any) => trainee.status.status !== 'drop',
      );
      setAllTrainees(sanitizedTrainees);
    }
  }, [trainees]);
  useEffect(() => {
    setFilteredTrainees(
      allTrainees.filter((trainee) =>
        trainee.profile.name.toLowerCase().includes(searchName.toLowerCase()),
      ),
    );
  }, [allTrainees, searchName]);

  if (!isVisible) return null;

  const handleGiveAttendance = (
    name: string,
    score: number,
    id: string,
    day: string,
  ) => {
    let updatedAttendance;
    const updateAttendance = traineesAttendance.find(
      (attendance) => attendance.id === id,
    );
    if (updateAttendance) {
      updatedAttendance = traineesAttendance.map((attendance) => {
        if (attendance.id === id) {
          return { ...attendance, score };
        }
        return attendance;
      });
    } else {
      updatedAttendance = [...traineesAttendance, { name, score, id, day }];
    }

    setTraineesAttendance(updatedAttendance);
  };

  const handleSubmitAttendance = () => {
    if (allTrainees.length !== traineesAttendance.length) {
      toast.error(
        'Give all trainees an attendance score before submitting attendance',
      );
      return;
    }

    const updatedRecords = traineesAttendance.map((attendance) => ({
      trainee: attendance.id,
      status: {
        day: attendance.day,
        score: attendance.score.toString(),
      },
    }));

    setRecordTrainees(updatedRecords);
  };

  return (
    <div className="fixed inset-0 bg-gray-900 bg-opacity-75 flex justify-center items-center z-50 p-4">
      <div className="bg-[#FFFFFF] dark:bg-[#020917] md:w-[500px] rounded-lg w-96 px-5 py-4 xmd:p-6 shadow-lg flex flex-col gap-3 text-[#111827] dark:text-white">
        <div className="w-full">
          <div className="w-full flex flex-col gap-3">
            <h2 className="text-[1.1rem] xmd:text-xl font-bold text-center text-purple-400 mb-2">
              Take Attendance
            </h2>
            <div className="flex justify-between">
              <h3 className="font-semibold text-[.85rem] xmd:text-[.95rem] ">
                Week {week}
              </h3>
              <p className="text-sm">
                {format(new Date(date), 'EE, do MMMM yyyy')}
              </p>
            </div>

            <div className="flex flex-col gap-2">
              <p className="font-semibold text-[..84rem] xmd:text-[.95rem]">
                Trainees:
              </p>
              <div className="w-full min-h-[40px] max-h-[110px] overflow-auto custom-scrollbar">
                {traineesAttendance.length ? (
                  <div className="grid grid-cols-2 gap-1 xmd:gap-3 justify-between">
                    {traineesAttendance.map((trainee) => (
                      <div
                        key={trainee.id}
                        className="flex justify-between items-center bg-gray-200 dark:bg-gray-700 rounded-[5px] text-[.8rem] xmd:text-sm px-3 py-[6px]"
                      >
                        <span>
                          {
                            // eslint-disable-next-line no-nested-ternary
                            trainee.name.length > 8 && window.innerWidth < 550
                              ? `${trainee.name.slice(0, 8)}...`
                              : window.innerWidth > 550 &&
                                trainee.name.length > 15
                              ? `${trainee.name.slice(0, 13)}...`
                              : trainee.name
                          }
                        </span>
                        <div className="flex items-center gap-2">
                          <span>[{trainee.score}]</span>
                          {trainee.score === 0 && (
                            <AttendanceSymbols status={0} />
                          )}
                          {trainee.score === 1 && (
                            <AttendanceSymbols status={1} />
                          )}
                          {trainee.score === 2 && (
                            <AttendanceSymbols status={2} />
                          )}
                        </div>
                      </div>
                    ))}
                  </div>
                ) : (
                  <p className="text-center text-[.83rem] xmd:text-[.9rem] font-medium">
                    No trainee given attendance.
                  </p>
                )}
              </div>
            </div>

            <div className="mt-4">
              <label className="block text-[.86rem] xmd:text-[.95rem] font-medium">
                Trainee Name
              </label>
              <div className="flex">
                <div className="flex-1 flex flex-col gap-1">
                  <input
                    type="text"
                    value={searchName}
                    onFocus={() => setInputFocus(true)}
                    onBlur={() => setInputFocus(false)}
                    onChange={(e) => {
                      setSearchName(e.target.value);
                    }}
                    className="placeholder:text-neutral-600 dark:placeholder:text-neutral-400 text-[.82rem] xmd:text-sm bg-gray-200 dark:bg-gray-700 w-full py-[5px] px-3 rounded-[2px] focus:outline-none focus:ring-1 focus:ring-purple-500"
                    placeholder="Find trainee by name"
                  />

                  <div className="min-h-[100px]">
                    {inputFocus && (
                      <div className="bg-[#EDF2EE] dark:bg-[#202B3E] z-20 left-0 right-0 w-full max-h-[100px] overflow-y-auto p-2 flex flex-col gap-1 custom-scrollbar">
                        {filteredTrainees.length === 0 ? (
                          <p className="text-[.78rem] xmd:text-sm">{`No trainee found in team ${teamName}.`}</p>
                        ) : (
                          filteredTrainees.map((trainee) => (
                            <div
                              key={trainee.email}
                              onMouseDown={(e) => e.preventDefault()}
                              className="flex items-center border-[#858585] border-b-[1px] py-[2px] xmd:py-1"
                            >
                              <p className="flex-1 text-[.83rem] xmd:text-sm">
                                {trainee.profile.name}
                              </p>
                              <div className="flex items-center gap-1 xmd:gap-2">
                                <div
                                  onMouseDown={(e) => {
                                    e.preventDefault();
                                    handleGiveAttendance(
                                      trainee.profile.name,
                                      2,
                                      trainee.id,
                                      day,
                                    );
                                  }}
                                  className="cursor-pointer hover:brightness-75"
                                  data-testid="test-score-2"
                                >
                                  <AttendanceSymbols status={2} />
                                </div>
                                <div
                                  onMouseDown={(e) => {
                                    e.preventDefault();
                                    handleGiveAttendance(
                                      trainee.profile.name,
                                      1,
                                      trainee.id,
                                      day,
                                    );
                                  }}
                                  className="cursor-pointer hover:brightness-75"
                                  data-testid="test-score-1"
                                >
                                  <AttendanceSymbols status={1} />
                                </div>
                                <div
                                  onMouseDown={(e) => {
                                    e.preventDefault();
                                    handleGiveAttendance(
                                      trainee.profile.name,
                                      0,
                                      trainee.id,
                                      day,
                                    );
                                  }}
                                  className="cursor-pointer hover:brightness-75"
                                  data-testid="test-score-0"
                                >
                                  <AttendanceSymbols status={0} />
                                </div>
                              </div>
                            </div>
                          ))
                        )}
                      </div>
                    )}
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div className="mt-3 flex justify-between text-[.9rem] xmd:text-[.95rem]">
          <button
            type="button"
            onClick={() => {
              setTraineesAttendance([]);
              onClose();
            }}
            className="bg-gray-500 text-white py-[4px] xmd:py-[6px] px-4 rounded-[4px] hover:bg-gray-600"
          >
            Cancel
          </button>
          <button
            type="submit"
            disabled={traineesAttendance.length === 0}
            className={`py-[4px] xmd:py-[6px] px-4 rounded-[4px] ${
              traineesAttendance.length === 0 || loadingRecordAttendance
                ? 'bg-gray-300 text-gray-800 cursor-not-allowed'
                : 'bg-purple-500 text-white hover:bg-purple-600'
            }`}
            onClick={() => {
              handleSubmitAttendance();
            }}
          >
            {loadingRecordAttendance ? 'Loading...' : 'Submit'}
          </button>
        </div>
      </div>
    </div>
  );
}

export default ModalAttendance;
