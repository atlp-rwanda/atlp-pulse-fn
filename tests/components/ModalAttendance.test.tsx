import React, { useState } from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { InMemoryCache } from '@apollo/client';
import ModalAttendance from '../../src/components/ModalAttendance';
import { RECORD_ATTENDANCE } from '../../src/Mutations/Attendance';

const closeModal = jest.fn();
const setTraineeAttendanceDataMock = jest.fn();


const cache = new InMemoryCache();

const sampleTrainees = {
  getTeamTrainees: [
    {
      id: 'trainee-123',
      email: 'test-trainee-name@gmail.com',
      profile: {
        name: 'test-trainee-name',
      },
      status: {
        status: 'active',
      },
    },
    {
      id: 'trainee-456',
      email: 'test-trainee-name2@gmail.com',
      profile: {
        name: 'test-trainee-name2',
      },
      status: {
        status: 'active',
      },
    },
  ],
};

const mocks = [
  {
    request: {
      query: RECORD_ATTENDANCE,
      variables: {
        week: 3,
        team: 'Team-I-id-123',
        date: '11-11-2024',
        trainees: [
          {
            trainee: 'trainee-123',
            status: {
              day: 'mon',
              score: '0',
            },
          },
          {
            trainee: 'trainee-456',
            status: {
              day: 'mon',
              score: '1',
            },
          },
        ],
        orgToken: 'mocked-org-token',
      },
    },
    result: {
      data: {
        recordAttendance: {
          team: {
            id: 'Team-I-id-123',
            name: 'Team I',
            cohort: {
              name: 'Cohort 1',
            },
          },
          trainees: [
            {
              trainee: {
                profile: {
                  name: 'test-trainee-name',
                },
              },
              status: [
                {
                  day: 'mon',
                  date: '1727806679051',
                  score: '0',
                },
              ],
            },
            {
              trainee: {
                profile: {
                  name: 'test-trainee-name2',
                },
              },
              status: [
                {
                  day: 'mon',
                  date: '1727806679051',
                  score: '0',
                },
              ],
            },
          ],
        },
      },
    },
    maxUsageCount: 5,
  },
];
// eslint-disable-next-line no-undef
const errorMocks = [
  {
    request: {
      query: RECORD_ATTENDANCE,
    },
    variableMatcher: jest.fn().mockReturnValue(true),
    error: new Error("An error occurred"),
    maxUsageCount: 5,
  },
];

describe('Record attendance modal', () => {
  beforeEach(() => {
    localStorage.setItem('orgToken', 'mocked-org-token');
  });

  afterEach(() => {
    localStorage.clear();
    cleanup();
  });
  it('Renders Record attendance modal', () => {
    const elem = renderer
      .create(
        <MockedProvider>
          <ModalAttendance
            day="mon"
            week={3}
            isVisible
            onClose={closeModal}
            team="Team-I-id-123"
            date="11-11-2024"
            teamName="Team I"
            trainees={sampleTrainees}
            setAttendanceData={() => true}
          />
        </MockedProvider>,
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });
  it('Closes the modal', () => {
    render(
      <MockedProvider>
        <ModalAttendance
          day="mon"
          week={3}
          isVisible
          onClose={closeModal}
          team="Team-I-id-123"
          date="11-11-2024"
          teamName="Team I"
          trainees={sampleTrainees}
          setAttendanceData={() => true}
        />
      </MockedProvider>,
    );

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeInTheDocument();

    fireEvent.click(cancelButton);
  });
  it('Renders Record attendance modal and focus on an input field', async () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ModalAttendance
          day="mon"
          week={3}
          isVisible
          onClose={closeModal}
          team="Team-I-id-123"
          date="11-11-2024"
          teamName="Team I"
          trainees={sampleTrainees}
          setAttendanceData={() => true}
        />
      </MockedProvider>,
    );

    const inputElement = screen.getByRole('textbox');
    expect(inputElement).toBeInTheDocument();

    fireEvent.focus(inputElement);
    fireEvent.blur(inputElement);
    fireEvent.focus(inputElement);
    fireEvent.change(inputElement, { target: { value: 'test' } });

    const traineeRowData = screen.getByText('test-trainee-name');
    expect(traineeRowData).toBeInTheDocument();

    const testScore2 = screen.getAllByTestId('test-score-2');
    expect(testScore2[0]).toBeInTheDocument();

    fireEvent.mouseDown(testScore2[0]);

    const testScore1 = screen.getAllByTestId('test-score-1');
    expect(testScore1[0]).toBeInTheDocument();

    fireEvent.mouseDown(testScore1[0]);

    const testScore0 = screen.getAllByTestId('test-score-0');
    expect(testScore0[0]).toBeInTheDocument();

    fireEvent.mouseDown(testScore0[0]);

    const submitButton = screen.getByText('Submit');
    expect(submitButton).toBeInTheDocument();

    fireEvent.click(submitButton);

    fireEvent.mouseDown(testScore0[1]);

    fireEvent.mouseDown(testScore1[1]);

    fireEvent.click(submitButton);

    expect(await screen.findByText('Loading...')).toBeInTheDocument();
    expect(await screen.findByText('Submit')).toBeInTheDocument();
  });
  it('Renders Record attendance modal and return error on submission', async () => {
    render(
      <MockedProvider mocks={errorMocks} addTypename={false}>
        <ModalAttendance
          day="mon"
          week={3}
          isVisible
          onClose={closeModal}
          team="Team-I-id-123"
          date='11-11-2024'
          teamName="Team I"
          trainees={{ getTeamTrainees: [sampleTrainees.getTeamTrainees[0]] }}
          setAttendanceData={() => true}
        />
      </MockedProvider>,
    );

    const inputElement = screen.getByRole('textbox');
    expect(inputElement).toBeInTheDocument();

    fireEvent.focus(inputElement);
    fireEvent.change(inputElement, { target: { value: 't' } });

    const traineeRowData = screen.getByText('test-trainee-name');
    expect(traineeRowData).toBeInTheDocument();

    const testScore2 = screen.getAllByTestId('test-score-2');
    expect(testScore2[0]).toBeInTheDocument();

    fireEvent.mouseDown(testScore2[0]);

    const submitButton = screen.getByText('Submit');
    expect(submitButton).toBeInTheDocument();

    fireEvent.click(submitButton);

    expect(await screen.findByText('Loading...')).toBeInTheDocument();
    expect(await screen.findByText('Submit')).toBeInTheDocument();
  });
  it('Doesn\'t Record attendance modal, when prop isvisible has falsy value', async () => {
    render(
      <MockedProvider mocks={errorMocks} addTypename={false}>
        <ModalAttendance
          day="mon"
          week={3}
          isVisible={false}
          onClose={closeModal}
          team="Team-I-id-123"
          date='11-11-2024'
          teamName="Team I"
          trainees={{ getTeamTrainees: [sampleTrainees.getTeamTrainees[0]] }}
          setAttendanceData={() => true}
        />
      </MockedProvider>,
    );

    const submitButton = screen.queryByText('Submit');
    expect(submitButton).toBeNull();
  });
  it('Render Record attendance modal,attendance can\'t be made, there are no trainees in team', async () => {
    render(
      <MockedProvider mocks={errorMocks} addTypename={false}>
        <ModalAttendance
          day="mon"
          week={3}
          isVisible
          onClose={closeModal}
          team="Team-I-id-123"
          date="11-11-2024"
          teamName="Team I"
          trainees={{ getTeamTrainees: [sampleTrainees.getTeamTrainees[0]] }}
          setAttendanceData={() => true}
        />
      </MockedProvider>,
    );

    const inputElement = screen.getByRole('textbox');
    expect(inputElement).toBeInTheDocument();

    fireEvent.focus(inputElement);
    fireEvent.change(inputElement, { target: { value: 'test-xyz-abcde' } });

    expect(
      screen.getByText('No trainee found in team Team I.'),
    ).toBeInTheDocument();
  });
});
