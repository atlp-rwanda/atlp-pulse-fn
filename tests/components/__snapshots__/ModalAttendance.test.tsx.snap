// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Record attendance modal Renders Record attendance modal 1`] = `
<div
  className="fixed inset-0 bg-gray-900 bg-opacity-75 flex justify-center items-center z-50 p-4"
>
  <div
    className="bg-[#FFFFFF] dark:bg-[#020917] md:w-[500px] rounded-lg w-96 px-5 py-4 xmd:p-6 shadow-lg flex flex-col gap-3 text-[#111827] dark:text-white"
  >
    <div
      className="w-full"
    >
      <div
        className="w-full flex flex-col gap-3"
      >
        <h2
          className="text-[1.1rem] xmd:text-xl font-bold text-center text-purple-400 mb-2"
        >
          Take Attendance
        </h2>
        <div
          className="flex justify-between"
        >
          <h3
            className="font-semibold text-[.85rem] xmd:text-[.95rem] "
          >
            Week 
            3
          </h3>
          <p
            className="text-sm"
          >
            Mon, 11th November 2024
          </p>
        </div>
        <div
          className="flex flex-col gap-2"
        >
          <p
            className="font-semibold text-[..84rem] xmd:text-[.95rem]"
          >
            Trainees:
          </p>
          <div
            className="w-full min-h-[40px] max-h-[110px] overflow-auto custom-scrollbar"
          >
            <p
              className="text-center text-[.83rem] xmd:text-[.9rem] font-medium"
            >
              No trainee given attendance.
            </p>
          </div>
        </div>
        <div
          className="mt-4"
        >
          <label
            className="block text-[.86rem] xmd:text-[.95rem] font-medium"
          >
            Trainee Name
          </label>
          <div
            className="flex"
          >
            <div
              className="flex-1 flex flex-col gap-1"
            >
              <input
                className="placeholder:text-neutral-600 dark:placeholder:text-neutral-400 text-[.82rem] xmd:text-sm bg-gray-200 dark:bg-gray-700 w-full py-[5px] px-3 rounded-[2px] focus:outline-none focus:ring-1 focus:ring-purple-500"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Find trainee by name"
                type="text"
                value=""
              />
              <div
                className="min-h-[100px]"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="mt-3 flex justify-between text-[.9rem] xmd:text-[.95rem]"
    >
      <button
        className="bg-gray-500 text-white py-[4px] xmd:py-[6px] px-4 rounded-[4px] hover:bg-gray-600"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="py-[4px] xmd:py-[6px] px-4 rounded-[4px] bg-gray-300 text-gray-800 cursor-not-allowed"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;
